File: ios/BitcoinTaprootCoin.swift
import Foundation
import WalletCore

// BIP-86 key-path taproot (P2TR, bech32m `bc1p…`). Mirrors
// BITCOIN_ADDRESS_TYPES.bitcoin_taproot in
// dok-wallet-blockchain-networks/service/bitcoinHdAddress.js: purpose 86',
// plain xpub/xprv (tpub/tprv on testnet), WIF holds the untweaked key.
class BitcoinTaprootCoin: CoinFactory.Coin {

  override init(mnemonic: String) {
    super.init(mnemonic: mnemonic)
  }

  private func derivation(isTestNet: Bool) -> Derivation {
    return isTestNet ? .bitcoinTestnet : .bitcoinTaproot
  }

  // WalletCore tweaks the key (BIP-341) and bech32m-encodes it for mainnet
  // only. For testnet we re-encode the same witness program under the `tb`
  // hrp: Bech32.decodeM/encodeM convert the whole 5-bit symbol stream to
  // bytes and back, which is lossless for a v1 program (1 version symbol +
  // 52 program symbols = 265 bits = 33 bytes + one zero pad bit), so this
  // is a pure hrp/checksum swap.
  private func buildTaprootAddress(publicKey: PublicKey, isTestNet: Bool) -> String {
    let mainnetAddress = AnyAddress(publicKey: publicKey, coin: .bitcoin, derivation: .bitcoinTaproot).description
    guard isTestNet, let program = Bech32.decodeM(string: mainnetAddress) else {
      return mainnetAddress
    }
    return Bech32.encodeM(hrp: "tb", data: program)
  }

  private func firstReceiveKey(isTestNet: Bool) -> PrivateKey {
    return wallet.getKey(coin: .bitcoin, derivationPath: accountBasePath(isTestNet: isTestNet) + "/0/0")
  }

  override func getNewAddress(isTestNet: Bool) -> String {
    let publicKey = firstReceiveKey(isTestNet: isTestNet).getPublicKeySecp256k1(compressed: true)
    return buildTaprootAddress(publicKey: publicKey, isTestNet: isTestNet)
  }

  override func getPrivateKey(isTestNet: Bool) -> String {
    return Utils.convertToWif(data: firstReceiveKey(isTestNet: isTestNet).data, isTestNet: isTestNet, prefix: [0x80], testNetPrefix: [0xef])
  }

  override func getExtendedPublicKey(isTestNet: Bool) -> String {
    return wallet.getExtendedPublicKeyDerivation(purpose: .bip86, coin: .bitcoin, derivation: derivation(isTestNet: isTestNet), version: isTestNet ? .tpub : .xpub)
  }

  override func getExtendedPrivateKey(isTestNet: Bool) -> String {
    return wallet.getExtendedPrivateKeyDerivation(purpose: .bip86, coin: .bitcoin, derivation: derivation(isTestNet: isTestNet), version: isTestNet ? .tprv : .xprv)
  }

  override func signTransaction(rawData: String) -> String {
    return ""
  }

  override func addCustomDerivation(derivePath: String, isTestNet: Bool) -> NSMutableDictionary {
    let privateKey = wallet.getKey(coin: .bitcoin, derivationPath: derivePath)
    let publicKey = privateKey.getPublicKeySecp256k1(compressed: true)
    let dict: NSMutableDictionary = [:]
    dict["derivePath"] = derivePath
    dict["privateKey"] = Utils.convertToWif(data: privateKey.data, isTestNet: isTestNet, prefix: [0x80], testNetPrefix: [0xef])
    dict["address"] = buildTaprootAddress(publicKey: publicKey, isTestNet: isTestNet)
    return dict
  }

  override func accountBasePath(isTestNet: Bool) -> String {
    return isTestNet ? "m/86'/1'/0'" : "m/86'/0'/0'"
  }

  override func getDeriveAddresses(isTestNet: Bool) -> NSMutableArray {
    // BIP86 standard: 20 external/receive (…/0/i) + 20 internal/change (…/1/i)
    let result = NSMutableArray()
    result.addObjects(from: getDeriveAddressRange(chainIndex: 0, startIndex: 0, count: 20, isTestNet: isTestNet) as! [Any])
    result.addObjects(from: getDeriveAddressRange(chainIndex: 1, startIndex: 0, count: 20, isTestNet: isTestNet) as! [Any])
    return result
  }
}

File: android/app/src/main/java/com/coinswallet/coins/BitcoinTaprootCoin.java
// BitcoinTaprootCoin.java
package com.coinswallet.coins;

import com.coinswallet.CoinFactory;
import com.coinswallet.Utils;

import wallet.core.jni.AnyAddress;
import wallet.core.jni.Bech32;
import wallet.core.jni.Derivation;
import wallet.core.jni.HDVersion;
import wallet.core.jni.HDWallet;
import wallet.core.jni.CoinType;
import wallet.core.jni.PublicKey;
import wallet.core.jni.Purpose;
import com.facebook.react.bridge.Arguments;
import com.facebook.react.bridge.ReadableArray;
import com.facebook.react.bridge.ReadableMap;
import com.facebook.react.bridge.WritableArray;
import com.facebook.react.bridge.WritableMap;
import wallet.core.jni.PrivateKey;

// BIP-86 key-path taproot (P2TR, bech32m `bc1p…`). Mirrors
// BITCOIN_ADDRESS_TYPES.bitcoin_taproot in
// dok-wallet-blockchain-networks/service/bitcoinHdAddress.js: purpose 86',
// plain xpub/xprv (tpub/tprv on testnet), WIF holds the untweaked key.
public class BitcoinTaprootCoin extends CoinFactory.Coin {
    private final HDWallet wallet;
    byte[] prefix = new byte[]{(byte) 0x80};
    byte[] testnetPrefix = new byte[]{(byte) 0xef};

    public BitcoinTaprootCoin(String mnemonic) {
        super(mnemonic);
        this.wallet = super.wallet;
    }

    private Derivation derivationFor(Boolean isTestNet) {
        return Boolean.TRUE.equals(isTestNet) ? Derivation.BITCOINTESTNET : Derivation.BITCOINTAPROOT;
    }

    // wallet-core tweaks the key (BIP-341) and bech32m-encodes it for mainnet
    // only. For testnet we re-encode the same witness program under the `tb`
    // hrp: TWBech32 decodeM/encodeM convert the whole 5-bit symbol stream to
    // bytes and back, which is lossless for a v1 program (1 version symbol +
    // 52 program symbols = 265 bits = 33 bytes + one zero pad bit), so this
    // is a pure hrp/checksum swap.
    private String buildTaprootAddress(PublicKey publicKey, Boolean isTestNet) {
        String mainnetAddress = new AnyAddress(publicKey, CoinType.BITCOIN, Derivation.BITCOINTAPROOT).description();
        if (!Boolean.TRUE.equals(isTestNet)) {
            return mainnetAddress;
        }
        return Bech32.encodeM("tb", Bech32.decodeM(mainnetAddress));
    }

    private PrivateKey firstReceiveKey(Boolean isTestNet) {
        return wallet.getKey(CoinType.BITCOIN, accountBasePath(isTestNet) + "/0/0");
    }

    @Override
    public String getNewAddress(Boolean isTestNet) {
        PublicKey publicKey = firstReceiveKey(isTestNet).getPublicKeySecp256k1(true);
        return buildTaprootAddress(publicKey, isTestNet);
    }

    @Override
    public String getPrivateKey(Boolean isTestNet) {
        return Utils.convertPrivateKeytoWIF(firstReceiveKey(isTestNet).data(), isTestNet, prefix, testnetPrefix);
    }

    @Override
    public String getExtendedPublicKey(Boolean isTestNet) {
        HDVersion version = Boolean.TRUE.equals(isTestNet) ? HDVersion.TPUB : HDVersion.XPUB;
        return wallet.getExtendedPublicKeyDerivation(Purpose.BIP86, CoinType.BITCOIN, derivationFor(isTestNet), version);
    }

    @Override
    public String getExtendedPrivateKey(Boolean isTestNet) {
        HDVersion version = Boolean.TRUE.equals(isTestNet) ? HDVersion.TPRV : HDVersion.XPRV;
        return wallet.getExtendedPrivateKeyDerivation(Purpose.BIP86, CoinType.BITCOIN, derivationFor(isTestNet), version);
    }

    @Override
    public String signTransaction(String rawData) {
        return null;
    }

    @Override
    public ReadableMap addCustomDerivation(String derivePath, Boolean isTestNet) {
        WritableMap obj = Arguments.createMap();
        PrivateKey tempPrivateKey = wallet.getKey(CoinType.BITCOIN, derivePath);
        PublicKey publicKey = tempPrivateKey.getPublicKeySecp256k1(true);
        String address = buildTaprootAddress(publicKey, isTestNet);
        String privateKeyString = Utils.convertPrivateKeytoWIF(tempPrivateKey.data(), isTestNet, prefix, testnetPrefix);
        obj.putString("address", address);
        obj.putString("derivePath", derivePath);
        obj.putString("privateKey", privateKeyString);
        return obj;
    }

    @Override
    public String accountBasePath(Boolean isTestNet) {
        return Boolean.TRUE.equals(isTestNet) ? "m/86'/1'/0'" : "m/86'/0'/0'";
    }

    @Override
    public ReadableArray getDeriveAddresses(Boolean isTestNet) {
        // BIP86 standard: 20 external/receive (…/0/i) + 20 internal/change (…/1/i)
        WritableArray result = Arguments.createArray();
        appendDeriveAddressRange(result, 0, 0, 20, isTestNet);
        appendDeriveAddressRange(result, 1, 0, 20, isTestNet);
        return result;
    }
}
